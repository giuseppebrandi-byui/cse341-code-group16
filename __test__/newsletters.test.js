const dotenv = require('dotenv');
dotenv.config();
const { MongoClient } = require('mongodb');
const request = require("supertest");
const { app } = require("../server");
const mongodb = require('../data/database');

describe('Newsletters API Tests', () => {
  let mongoClient;
  let db;
  let server;
  const URI = process.env.MONGODB_URI;

  beforeAll(async () => {
    // Connect to the in-memory database
    mongoClient = await MongoClient.connect(URI, {
      useNewUrlParser: true,
      useUnifiedTopology: true,
    });
    db = mongoClient.db();

    // Replace the database connection in your app with the test database
    jest.spyOn(mongodb, 'getDatabase').mockImplementation(() => ({
      db: () => db
    }));

    // Start the server and store the reference
    server = app.listen(3001);
  });

  afterAll(async () => {
    // Clean up resources
    if (mongoClient) {
      await mongoClient.close();
    }
    if (server) {
      await new Promise((resolve) => server.close(resolve));
    }
    jest.restoreAllMocks();
  });

  let randomId;

  it("should get all newsletters", async () => {
    const response = await request(app).get("/newsletters");
    expect(response.statusCode).toBe(200);
    expect(Array.isArray(response.body)).toBe(true);
    expect(Object.keys(response.body[0])).toEqual(expect.arrayContaining(['_id', 'name', 'email']));
    randomId = response.body[0]['_id'] ?? '';
  });

  it("should get specific newsletters", async () => {
    const response = await request(app).get("/newsletters/" + randomId);
    expect(response.statusCode).toBe(200);
    expect(Object.keys(response.body)).toEqual(expect.arrayContaining(['_id', 'name', 'email']));
  });

});

