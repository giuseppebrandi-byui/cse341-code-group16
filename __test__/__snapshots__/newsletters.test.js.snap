// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Newsletters API Tests - GET Endpoints GET /newsletters should get all newsletters and match snapshot 1`] = `
[
  {
    "_id": "679e81faa04d19f37756ea87",
    "email": "john.doe@example.com",
    "name": "John Doe",
  },
  {
    "_id": "679e81faa04d19f37756ea88",
    "email": "michael.johnson@example.com",
    "name": "Michael Johnson",
  },
]
`;

exports[`Newsletters API Tests - GET Endpoints GET /newsletters should return 400 when no newsletters exist 1`] = `
{
  "error": {
    "message": "No newsletters subscriber found.",
  },
}
`;

exports[`Newsletters API Tests - GET Endpoints GET /newsletters/:id should get specific newsletter and match snapshot 1`] = `
{
  "_id": "679e81faa04d19f37756ea87",
  "email": "john.doe@example.com",
  "name": "John Doe",
}
`;

exports[`Newsletters API Tests - GET Endpoints GET /newsletters/:id should return 400 for invalid id format 1`] = `
{
  "error": {
    "message": "Please enter a valid id with a string of 24 hex characters.",
  },
}
`;

exports[`Newsletters API Tests - GET Endpoints GET /newsletters/:id should return 400 for non-existent id 1`] = `
{
  "error": {
    "message": "No newsletter subscriber found with entered id.",
  },
}
`;
